src/main/scala/com/advancedtelematic/campaigner/daemon/DeviceInstallationReportListener.scala

```diff
@@ -40,4 +45,24 @@ class DeviceInstallationReportListener()(implicit db: Database, ec: ExecutionCon
         _log.info(s"Got DeviceUpdateReport for device ${msg.device.show} which is not scheduled by campaigner, ignoring this message.")
     }
   }
+
+  private def addDeviceToFailedGroup(msg: DeviceInstallationReport): Unit =
```

Comment:
The code works perfectly fine, but the name of the function might be a bit misleading. Let's try to come up with a better one by writing down what does the function do. It takes a device installation report and:

- checks that the installation failed
- extracts the failure code
- extracts the campaign id
- fetches the failed device group _or_ creates one if nothing is found
- finally, adds a device to the group (actually doing what is stated in the name)

Whew, that's a list! So, from here we can go two ways: either create the most specific name that won't mislead anybody (for example, `parseInstallationReportAndAddDeviceToANewOrExistingFailedGroupIfItIsFailed`) or use a generic name so that it won't disclose any specifics (for example, `processInstallationReport`). I personally find neither of those options to be fine.

What else we can do? Let's take a look at the code around. One thing that strucks me is that we already parse the report in `apply` method, processing it differently based on the correlation ID. That might be a good place for us to step in and add a processing step for a failed device. Another thing would be to split the `addDeviceToFailedGroup` function into a smaller methods that compose well. For example, one method could perform the FDG lookup and creation, another one will add the device to the group returned by the previous method. The latter could be used in `apply` directly. Yet another idea is to pass to these utility method only the minimum of necessary information and not drag the whole report around. Following these steps, we can come up with a code that looks similar to this one:

```scala
  def apply(msg: DeviceInstallationReport): Future[Unit] = {
    val resultStatus = if (msg.result.success)
      DeviceStatus.successful
    else
      DeviceStatus.failed

    val f = msg.correlationId match {
      case CampaignCorrelationId(uuid) =>
        val campaignId = CampaignId(uuid)
        val addDeviceToFailedDeviceGroupFuture =
          addDeviceToFailedDeviceGroup(msg.namespace, campaignId, msg.result.code, msg.device)
        for {
          _ <- if (!msg.result.success) addDeviceToFailedDeviceGroupFuture else Future.unit
          _ <- campaigns.finishDevices(CampaignId(uuid), Seq(msg.device), resultStatus)
        } yield ()
      case MultiTargetUpdateId(uuid) => for {
        update <- updateRepo.findByExternalId(msg.namespace, ExternalUpdateId(uuid.toString))
        _ <- campaigns.finishDevice(update.uuid, msg.device, resultStatus)
      } yield ()
    }

    f.recover {
      case Errors.MissingExternalUpdate(_) =>
        _log.info(s"Could not find an update with external id ${msg.correlationId}, ignoring message")
      case Errors.DeviceNotScheduled =>
        _log.info(s"Got DeviceUpdateReport for device ${msg.device.show} which is not scheduled by campaigner, ignoring this message.")
    }
  }

  private def addDeviceToFailedDeviceGroup(namespace: Namespace, campaignId: CampaignId, failureCode: String, deviceId: DeviceId): Future[Unit] =
    findOrCreateFailedDeviceGroup(namespace, campaignId, failureCode)
      .flatMap(gid => deviceRegistryClient.addDeviceToGroup(namespace, gid, deviceId))

  private def findOrCreateFailedDeviceGroup(namespace: Namespace, campaignId: CampaignId, failureCode: String): Future[GroupId] =
    campaigns.fetchFailedGroupId(campaignId, failureCode).flatMap {
      case Some(g) =>
        FastFuture.successful(g)
      case None =>
        deviceRegistryClient
          .createGroup(namespace, s"failed-group-campaignId-$campaignId-failureCode-$failureCode")
          .flatMap(gid => campaigns.persistFailedGroup(campaignId, gid, failureCode))
    }
```

Now we are (hopefully) more clear about our intentions and :drum: :drum: :drum: use less code!

Reply:
Of course, if you believe that the code above could be improved even further, then go for it!

Reply:
Thanks, I like that suggestion :) I've added that code with some minor tweaks.